### PR TITLE
L1T phase-2: particle flow and correlator updates, including emulators

### PR DESCRIPTION
**Pull request by cecilecaillol**
#### PR description:

This is another PR to port all L1T Phase-2 developments to cms-sw. This PR targets the particle flow code. It includes several PRs made locally, the details of which can be found in the twiki https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideL1TPhase2Instructions and in the cms-l1t-offline branch.

### CODE REVIEW

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
+l1

**Comment by cecilecaillol** [comment]
@srimanob Can you please review and eventually sign off?

File: L1Trigger/Phase2L1ParticleFlow/plugins/DeregionizerProducer.cc [hunk 1]
```diff
@@ -0,0 +1,178 @@
+#include <unordered_map>
+
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "DataFormats/L1TParticleFlow/interface/PFCandidate.h"
+
+#include "L1Trigger/Phase2L1ParticleFlow/src/newfirmware/deregionizer/deregionizer_input.h"
```
Comment (srimanob) [inline L1Trigger/Phase2L1ParticleFlow/plugins/DeregionizerProducer.cc:10]: I'm not sure if this structure follows CMSSW rule. Maybe @perrotta @qliphy @smuzaffar can suggest.

File: L1Trigger/Phase2L1ParticleFlow/interface/PUAlgoBase.h [hunk 2]
```diff
@@ -17,16 +17,27 @@ namespace l1tpf_impl {
...
+    virtual void runNeutralsPU(Region &r,
+                               std::vector<float> &z0,
+                               float npu,
+                               const std::vector<float> &globals) const = 0;
 
   protected:
     int debug_;
     float etaCharged_, vtxRes_;
     bool vtxAdaptiveCut_;
+    int NVtx_;
```
Comment (srimanob) [inline L1Trigger/Phase2L1ParticleFlow/interface/PUAlgoBase.h:40]: State class member with lower case, i.e. nVtx_;

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1MHtPFProducer.cc [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+class L1MhtPfProducer : public edm::global::EDProducer<> {
+public:
+  explicit L1MhtPfProducer(const edm::ParameterSet&);
+  ~L1MhtPfProducer() override;
+
+private:
+  void produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const override;
+  edm::EDGetTokenT<std::vector<l1t::PFJet>> _jetsToken;
+  float _minJetPt;
```
Comment (srimanob) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1MHtPFProducer.cc:28]: Following the CMSSW naming code rule, we should not start class member with "_".

https://cms-sw.github.io/cms_coding_rules.html#2--naming-rules-1

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1MetPfProducer.cc [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+      4) Check px=-|px|.
+    */
+
+  // set phi to first quadrant
+  phi_t phiQ1 = (phi > 0) ? phi : phi_t(-phi);  // Q1/Q4
+  if (phiQ1 >= hwPiOverTwo_)
+    phiQ1 = hwPi_ - phiQ1;
+
+  if (phiQ1 > hwPiOverTwo_) {
+    std::cout << "unexpected phi (high)" << std::endl;
```
Comment (srimanob) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1MetPfProducer.cc:133]: Why do you use `std::cout` here? Use MessageLogger.

Not sure I understand the logic. You check `phiQ1` first, then update the value on the above `if`. Then check and update again here?

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1MetPfProducer.cc [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    phiQ1 = hwPiOverTwo_;
+  } else if (phiQ1 < 0) {
+    std::cout << "unexpected phi (low)" << std::endl;
```
Comment (srimanob) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1MetPfProducer.cc:136]: same here for message logger.

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1SeedConePFJetProducer.cc [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+  explicit L1SeedConePFJetProducer(const edm::ParameterSet&);
+  ~L1SeedConePFJetProducer() override;
+
+private:
+  /// ///////////////// ///
+  /// MANDATORY METHODS ///
+  void produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const override;
+  /// ///////////////// ///
+
+  float _coneSize;
```
Comment (srimanob) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1SeedConePFJetProducer.cc:32]: Following the CMSSW code rule, we should not start the class member with "_".

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrectedPFJetProducer.cc [hunk 7]
```diff
@@ -1,5 +1,8 @@
 #include "DataFormats/L1TParticleFlow/interface/PFJet.h"
 #include "DataFormats/JetReco/interface/Jet.h"
+#include "DataFormats/Candidate/interface/Candidate.h"
+
+#include "DataFormats/Candidate/interface/Candidate.h"
```
Comment (srimanob) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrectedPFJetProducer.cc:5]: duplicated above line?

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrectedPFJetProducer.cc [hunk 8]
```diff
@@ -42,11 +45,11 @@ void L1TCorrectedPFJetProducer::produce(edm::StreamID, edm::Event& iEvent, const
     out->emplace_back(srcjet.p4());
     auto& jet = out->back();
     // copy daughters
-    if (copyDaughters_) {
-      for (const auto& dau : srcjet.daughterPtrVector()) {
-        jet.addConstituent(edm::Ptr<l1t::L1Candidate>(dau));
-      }
-    }
+    //if (copyDaughters_) {
```
Comment (srimanob) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrectedPFJetProducer.cc:48]: Do you still need it? If not, delete. Git keeps history already.

File: L1Trigger/Phase2L1ParticleFlow/plugins/DeregionizerProducer.cc (hunk 1 above)
Comment (perrotta) [reply L1Trigger/Phase2L1ParticleFlow/plugins/DeregionizerProducer.cc:10]: Uhm, no, it looks weird,
If the header in L1Trigger/Phase2L1ParticleFlow/src/newfirmware/deregionizer/deregionizer_input.h is needed by CMSSW code in plugins, why not to put it in  L1Trigger/Phase2L1ParticleFlow/interface? Or maybe inside the /plugin area itself, if it is not needed anywhere else (I did not look at the code).
Even worse the inclusions of the cpp files in the lines below: why do you need to include the implementation file, besides the header?

Comment (cecilecaillol) [reply L1Trigger/Phase2L1ParticleFlow/plugins/DeregionizerProducer.cc:10]: About the inclusion of the cpp files, it was done to solve compilation errors "undefined references to XXX".
I am not sure I understand the suggestion. Should I change the structure of the code and move the files in src/newfirmware/deregionizer somewhere else, or should I modify where the header files are included?

Comment (perrotta) [reply L1Trigger/Phase2L1ParticleFlow/plugins/DeregionizerProducer.cc:10]: > About the inclusion of the cpp files, it was done to solve compilation errors "undefined references to XXX".

Only visible interfaces should stay in the .h files. If anything needed from the outside world is found in the implementation file, then is should be moved to the interface

> I am not sure I understand the suggestion. Should I change the structure of the code and move the files in src/newfirmware/deregionizer somewhere else, or should I modify where the header files are included?

If deregionizer_input.h needs only to be accessed from the  L1Trigger/Phase2L1ParticleFlow/plugin area, then move it over there

If it needs to be accessed even from outside that plugin area, then move to L1Trigger/Phase2L1ParticleFlow/interface

By the way, which is the purpose of the L1Trigger/Phase2L1ParticleFlow/src/newfirmware/deregionizer area? (I repeat myself, I did not look at the code, yet). Is there any reason to move it away from main L1Trigger/Phase2L1ParticleFlow? If it acts as a separate package, then you should rather develop it in an independent (new?) package

Comment (cecilecaillol) [reply L1Trigger/Phase2L1ParticleFlow/plugins/DeregionizerProducer.cc:10]: @gpetruc Can you please comment about the structure of L1Trigger/Phase2L1ParticleFlow/src/newfirmware and whether we can reorganize the code to respect CMS standards?

**Comment by gpetruc** [comment]
Hi,

The code in the newfirmware directory is that way because the very same
code exists also as standalone for compilation outside of CMSSW and inside
Vivado when building and testing the firmware. That's why it doesn't follow
well the CMS conventions and layout.
Originally we planned to move it to an external, but now we're reorganizing
the firmware repository to take some source from CMSSW and so we will
eventually be able to reorganize things and make it more CMSSW-conformant
(but it may take a month or two)

Il Ven 8 Apr 2022, 10:08 cecilecaillol ***@***.***> ha
scritto:

**Comment by cecilecaillol** [comment]
please test

File: L1Trigger/Phase2L1ParticleFlow/plugins/DeregionizerProducer.cc (hunk 1 above)
Comment (gpetruc) [reply L1Trigger/Phase2L1ParticleFlow/plugins/DeregionizerProducer.cc:10]: Regarding the inclusion of the cpp files, it wasn't of my doing: in 11_1_X, the cpp files in the subdirectories were being properly compiled by scram so there was no need to include them. 
I don't know if something was changed in the BuildFile syntax in how scram works in the newer release.

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc [hunk 9]
```diff
@@ -0,0 +1,1107 @@
...
+  float z0 = 0;
+  double ptsum = 0;
+  l1t::VertexWord pvwd;
+  // FIXME: collections seem to be already sorted
+  if (emuTkVtx_) {
+    edm::Handle<std::vector<l1t::VertexWord>> vtxEmuHandle;
+    iEvent.getByToken(tkVtxEmu_, vtxEmuHandle);
+    for (const auto &vtx : *vtxEmuHandle) {
+      ptsums.push_back(std::pair<float, float>(vtx.pt(), vtx.z0()));
+      if (ptsum == 0 || vtx.pt() > ptsum) {
```
Comment (perrotta) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc:369]: `ptsum` and `z0` are not used here: either a line like L385 below is missing here, or maybe that L385 shoud have been moved out of the `if` block

File: L1Trigger/Phase2L1ParticleFlow/src/PUAlgoBase.cc [hunk 10]
```diff
@@ -75,6 +95,109 @@ void PUAlgoBase::doVertexing(std::vector<Region> &rs, VertexAlgo algo, float &pv
...
+          bool pPass = false;
+          for (int v = 0; v < vtx; v++) {
+            if (lBin[v] == b)
+              pPass = true;
+          }
+          if (pPass)
+            continue;
+          float pVal = h_dz->GetBinContent(b);
+          if (pMax < pVal || lBin[vtx] == -1) {
+            pVal = pMax;
```
Comment (perrotta) [inline L1Trigger/Phase2L1ParticleFlow/src/PUAlgoBase.cc:150]: Why do you need to update `pVal` here?

**Review by perrotta — COMMENTED** [review]
This is not a full review: just reporting about a couple of issues pointed out by the static analyzer

**Comment by perrotta** [comment]
> @perrotta Given Giovanni's comment, can we leave the newfirmware directory structure like this for now, and open an issue to follow up within a month or two?

Thank you @cecilecaillol and @gpetruc.

However, I think it would be a bad practice integrating badly organized code (which interacts with our current L1 code and requires non-trivial modifications to it). Given also the huge complexity of this PR as a whole, can I suggest a procedure to integrate it, which should help and simplify the work of both the authors and the reviewer?

My suggestion is to factorize this PR in two pieces:
- PR nr 1 moves all is now hanging here in L1Trigger/Phase2L1ParticleFlow/src/newfirmware into a new package under L1Trigger. Further updates and modifications to it can be done at a separate time, when you are ready to it. This shouldn't take that much time to implement.
- PR nr 2 integrates the rest of this PR with the correct references to the new package merged with the other PR.

This would make the review of the code easier (although still complicate due to its size). Not only for the external reviewer, but also for you: the couple of issues (quite likely bugs) [reported by the static analyzer](https://github.com/cms-sw/cmssw/pull/37452#pullrequestreview-937219163) witness the presence of inaccuracies hidden in the code of the this PR that should better sorted out as much as possible before merging.

**Comment by gpetruc** [comment]
Hi Andrea,

As you can see from the code, essentially all classes in the
src/newfirmware are used individually from the main edproducers and
configured from sub PSets of the main Cff files, so moving all files in a
different place will make the two packages completely intertwined IMHO and
won't improve anything in the clarity or the reviewability of the code.

Now, any big reorganization of the code would have to wait for after the
P2UG review in mid May, but what I could offer to do before that if you
think it will help significantly with the integration is:
1) Move the firmware and low-level emulator data formats from
L1Trigger/Phase2L1ParticleFlow/src/newfirmware/dataformats to
DataFormats/L1TParticleFlow (this is something we want to do anyway
eventually, though later some formats which belong to GT and not Correlator
should go elsewhere, to be discussed with GT developers)
2) Remove "/newfirmware/" from the paths, but still keep files in different
algorithm  subdirectories otherwise the main
L1Trigger/Phase2L1ParticleFlow/src  directory would become a mess
3) remove more obsolete code from the packet

Giovanni



Il Dom 10 Apr 2022, 09:37 Andrea Perrotta ***@***.***> ha
scritto:

**Comment by perrotta** [comment]
Thank you Giovanni, I agree with your plan. 

Can I suggest further to separate the interface and the implementation files of the classess currrently in `newfimware`, letting hang all the interfaces from subdirectories defined in L1Trigger/Phase2L1ParticleFlow/interface, and the implementations from same name subdirectories in L1Trigger/Phase2L1ParticleFlow/src?
You can find examples of it, still in the L1Trigger area, in L1Trigger/L1TMuonOverlapPhase1 (e.g.)

**Comment by cecilecaillol** [comment]
I have opened a new PR with the reorganized code https://github.com/cms-sw/cmssw/pull/37536

**Comment by perrotta** [comment]
> I have opened a new PR with the reorganized code #37536

Thank you @cecilecaillol

**Pull request closed without merging**